spark-operator-docs/spark_application.md

```diff
@@ -0,0 +1,212 @@
+debug actual failure from the driver pods. Spark Operator would stage the last observed 
+driver pod status with the stopping state for audit purposes. 
+
+## Configure the Tolerations for SparkApplication
+
+### Restart
+
+Spark Operator enables configure app restart behavior for different failure types. Here's a 
+sample restart config snippet: 
+
+``` yaml
+restartConfig:
+  # accptable values are 'NEVER', 'ALWAYS', ON_FAILURE and 'ON_INFRASTRUCTURE_FAILURE'
+  restartPolicy: NEVER
+  # operator would retry the application if configured. All resources from current attepmt 
+  # would be deleted before starting next attempt
+  maxRestartAttempts: 3 
+  # backoff time (in millis) that operator would wait before next attempt
+  restartBackoffMillis: 30000
+```
+
+### Timeouts
+
+Example for configure timeouts:
+
+```yaml
+applicationTimeoutConfig:
+  # timeouts set to 5min
+
+  # time to wait for driver reaches running state after requested driver
+  driverStartTimeoutMillis: 300000
+
+  # time to wait for driver reaches ready state  
+  sparkSessionStartTimeoutMillis: 300000
+
+  # time to wait for driver to acquire minimal number of running executors
+  executorStartTimeoutMillis: 300000
+
+  # time to wait for force delete resources at the end of attempt
+  forceTerminationGracePeriodMillis: 300000
+```
+ 
+### Instance Config
+
+Instance Config helps operator to decide whether an application is running healthy. When 
+the underlying cluster has batch scheduler enabled, you may configure the apps to be 
+started if and only if there are sufficient resources. If, however, the cluster does not 
+have a batch scheduler, operator may help avoid app hanging with `InstanceConfig` that 
+describes the bare minimal tolerable scenario.
+
+For example, with below spec:
+
+```yaml
+applicationTolerations:
+  instanceConfig:
+    minExecutors: 3
+    initExecutors: 5
+    maxExecutors: 10
+sparkConf:
+  spark.executor.instances: "10"
+```
+
+Spark would try to bring up 10 executors as defined in SparkConf. In addition, from 
+operator perspective,
+
+* If Spark app acquires less than 5 executors in given tine window (.spec.
+  applicationTolerations.applicationTimeoutConfig.executorStartTimeoutMillis) after 
+  submitted, it would be shut down proactively in order to avoid resource deadlock. 
+* Spark app would be marked as 'RUNNING_WITH_PARTIAL_CAPACITY' if it loses executors after 
+  successfully start up.
+* Spark app would be marked as 'RUNNING_HEALTHY' if it has at least min executors after 
+  successfully started up.
+
+### Delete Resources On Termination
+
+Operator by default would delete all created resources at the end of an attempt. It would 
+try to record the last observed driver status in `status` field of the application for 
+troubleshooting purpose. 
+
+On the other hand, when developing an application, it's possible to configure
+
+```yaml
+applicationTolerations:
+  deleteOnTermination: false
```

Comment:
A common case: a user might want to delete the normal exited pod automatically but retain the abnormal exited pod for further diagnostic.

Reply:
love this idea! 

https://github.com/apache/spark-kubernetes-operator/pull/2/commits/91bc916728d8a820774233c4eef7260092d427c3